### PR TITLE
[node] Handle multi-protocol in Edge Functions

### LINKED ISSUE #9501: Vercel dev cannot handle requests to Edge functions with CloudFlare Tunnel
### Discussed in https://github.com/vercel/vercel/discussions/9500

<div type='discussions-op-text'>

<sup>Originally posted by **PonomareVlad** February 22, 2023</sup>
### Summary

Vercel CLI outputs this error on every request:
```
Unhandled rejection: Failed to parse URL from https,http://***.trycloudflare.com/: Invalid URL: https,http://***.trycloudflare.com/
```

Where `***` is some randomly generated domain from [CloudFlare Tunnel](https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/do-more-with-tunnels/trycloudflare/)

### Additional information

Steps to reproduce:
1. Install Vercel CLI `28.12.4`
2. Install [CloudFlare Tunnel](https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/do-more-with-tunnels/trycloudflare/) (`cloudflared`)
3. Make any [Edge function](https://vercel.com/docs/concepts/functions/edge-functions) ([or get example from repo](https://github.com/PonomareVlad/lit-ssr-vercel/tree/edge))
4. Run `vercel dev` (should work perfect on `localhost` origin)
5. Run `cloudflared tunnel --url http://localhost:3000`
6. After opening tunnel URL from console (ends with `.trycloudflare.com`)
7. You will see `FUNCTION_INVOCATION_FAILED` in browser
8. And this error in Vercel CLI output: `... Failed to parse URL from https,http://***.trycloudflare.com/ ...`

### Example

https://github.com/PonomareVlad/lit-ssr-vercel/tree/edge</div>

### PR DESCRIPTION
**Pull request by PonomareVlad**
This is proposal to fix #9501

### CODE REVIEW

File: packages/node/src/edge-functions/edge-handler-template.js [hunk 1]
```diff
@@ -5,7 +5,7 @@
 /* global IS_MIDDLEWARE ENTRYPOINT_LABEL */
 
 function buildUrl(requestDetails) {
-  let proto = requestDetails.headers['x-forwarded-proto'];
+  let proto = requestDetails.headers['x-forwarded-proto'].split(',').shift(); // handling multi-protocol like https,http://...
```
Comment (cb1kenobi) [inline packages/node/src/edge-functions/edge-handler-template.js]: After doing some Googling, indeed a comma separated list of protocols is valid. I have two concerns:

1. This code doesn't handle whitespace or non-word characters. Using a regex `/\b/` will split on word boundaries so if the header is set to something crazy like `https:` or `https http`, that will also work.

```suggestion
  let proto = requestDetails.headers['x-forwarded-proto'].split(/\b/).shift(); // handling multi-protocol like https,http://...
```

2. I'd love to see a test that demonstrates multiple protocols.

**Review by cb1kenobi — CHANGES_REQUESTED** [review]

File: packages/node/src/edge-functions/edge-handler-template.js (hunk 1 above)
Comment (PonomareVlad) [reply packages/node/src/edge-functions/edge-handler-template.js]: Thank you so much, now I'll try to add tests

Comment (PonomareVlad) [reply packages/node/src/edge-functions/edge-handler-template.js]: @cb1kenobi Sorry, I'm not sure for which package it is better to add tests: `cli`, `edge` or `node` ?

Comment (PonomareVlad) [reply packages/node/src/edge-functions/edge-handler-template.js]: In addition, I can’t successfully execute the `pnpm bootstrap` command, its always fails with error:
```
lerna ERR! EWORKSPACES Bootstrapping with pnpm is not supported. Use pnpm directly to manage dependencies: https://pnpm.io/cli/install
```
Снимок экрана 2023-02-23 в 01 30 20

The error looks the same as on my system and on Github Codespaces.

My versions: MacOS `13.2.1`, Node `18.13.0`, pnpm `7.27.1`

Comment (cb1kenobi) [reply packages/node/src/edge-functions/edge-handler-template.js]: Usually we just run `pnpm i` instead of bootstrap.

Comment (PonomareVlad) [reply packages/node/src/edge-functions/edge-handler-template.js]: @cb1kenobi I need some help, because I do not have sufficient experience in developing Vercel 🙂

Do I understand correctly that I need to use the `createEdgeEventHandler` method from `edge-handler` to create the `Edge Runtime` instance, to which I can send `IncomingMessage` with URL containing a double protocol ?

```ts
import { createEdgeEventHandler } from '../src/edge-functions/edge-handler';
import { IncomingMessage } from 'http';
import { Socket } from 'net';

test('Handle multi-protocol in Edge Functions', async () => {
  const handleEvent = await createEdgeEventHandler(/*...*/)
  const socket = new Socket()
  const req = new IncomingMessage(socket)
  req.url = "https,http://localhost/"
  const result = await handleEvent(req)
  expect(result.statusCode).toEqual(200)
})
```

Comment (cb1kenobi) [reply packages/node/src/edge-functions/edge-handler-template.js]: @PonomareVlad that is a great question! The short answer is I don't know. I mentioned this to the team and asked the best way to test this. I didn't get an absolute answer yet. I'll know more tomorrow.

Comment (EndangeredMassa) [reply packages/node/src/edge-functions/edge-handler-template.js]: I just opened a PR to allow at least unit-level tests of this logic. We can merge my PR first, then update this PR, and add some small tests for handling multiple protocols.
https://github.com/vercel/vercel/pull/9528

Comment (EndangeredMassa) [reply packages/node/src/edge-functions/edge-handler-template.js]: We had some fun CI issues, but the PR that makes this more testable landed. https://github.com/vercel/vercel/pull/9528

Now you can add a test for this behavior in `packages/node/test/unit/edge-functions/edge-handler-template.test.ts`.

Comment (PonomareVlad) [reply packages/node/src/edge-functions/edge-handler-template.js]: Awesome, thank you ! 
I will try do this tomorrow 🙂

**Review by cb1kenobi — APPROVED** [review]
LGTM

**Review by TooTallNate — APPROVED** [review]

**Review by cb1kenobi — APPROVED** [review]

**Comment by EndangeredMassa** [comment]
Thanks @PonomareVlad!
